File: swift/Sources/SmoothReading/SmoothToken.swift
import Foundation

/// A piece of the input: a word with its computed fixation split, or a separator.
/// Mirrors `Token` in `docs/SPEC.md` §4.
public enum SmoothToken: Sendable, Equatable {
    /// A word token.
    /// - Parameters:
    ///   - text: the whole word.
    ///   - fixation: length of the emphasised prefix, in grapheme clusters (`0` = not emphasised).
    ///   - fixationText: the emphasised prefix (empty when `fixation == 0`).
    ///   - restText: the remainder of the word.
    case word(text: String, fixation: Int, fixationText: String, restText: String)
    /// Anything that is not a word: whitespace, punctuation, HTML tags.
    case separator(text: String)

    /// The raw text of the token, so that concatenating all tokens rebuilds the input.
    public var text: String {
        switch self {
        case let .word(text, _, _, _): return text
        case let .separator(text): return text
        }
    }
}

File: android/smooth-reading-core/src/main/kotlin/io/smoothreading/SmoothToken.kt
package io.smoothreading

/** A piece of the input: either a word with a computed split, or a separator. */
public sealed interface SmoothToken {
    /** The slice of the input this token covers. */
    public val text: String

    /**
     * A word token (SPEC §4).
     *
     * @property fixation how many *grapheme clusters* are emphasised (`0` = none)
     * @property fixationText the emphasised prefix
     * @property restText everything after the prefix
     */
    public data class Word(
        override val text: String,
        val fixation: Int,
        val fixationText: String,
        val restText: String,
    ) : SmoothToken

    /** Whitespace, punctuation or markup: never emphasised. */
    public data class Separator(override val text: String) : SmoothToken
}
